src/Microsoft.AspNet.Razor/TagHelpers/TagHelperDescriptorComparer.cs

```diff
@@ -29,15 +30,28 @@ public class TagHelperDescriptorComparer : IEqualityComparer<TagHelperDescriptor
         /// <c>false</c> otherwise.</returns>
         /// <remarks>
         /// Determines equality based on <see cref="TagHelperDescriptor.TypeName"/>,
-        /// <see cref="TagHelperDescriptor.AssemblyName"/>, <see cref="TagHelperDescriptor.TagName"/> and
-        /// <see cref="TagHelperDescriptor.Prefix"/>.
+        /// <see cref="TagHelperDescriptor.AssemblyName"/>, <see cref="TagHelperDescriptor.TagName"/>,
+        /// <see cref="TagHelperDescriptor.Prefix"/>, <see cref="TagHelperDescriptor.Attributes"/>, and 
+        /// <see cref="TagHelperDescriptor.RequiredAttributes"/>.
         /// </remarks>
         public bool Equals(TagHelperDescriptor descriptorX, TagHelperDescriptor descriptorY)
         {
             return string.Equals(descriptorX.TypeName, descriptorY.TypeName, StringComparison.Ordinal) &&
                    string.Equals(descriptorX.TagName, descriptorY.TagName, StringComparison.OrdinalIgnoreCase) &&
                    string.Equals(descriptorX.Prefix, descriptorY.Prefix, StringComparison.OrdinalIgnoreCase) &&
-                   string.Equals(descriptorX.AssemblyName, descriptorY.AssemblyName, StringComparison.Ordinal);
+                   string.Equals(descriptorX.AssemblyName, descriptorY.AssemblyName, StringComparison.Ordinal) &&
+                   Enumerable.SequenceEqual(
+                       descriptorX.RequiredAttributes.OrderBy(
+                           attribute => attribute, StringComparer.OrdinalIgnoreCase),
+                       descriptorY.RequiredAttributes.OrderBy(
+                           attribute => attribute, StringComparer.OrdinalIgnoreCase),
+                       StringComparer.OrdinalIgnoreCase) &&
+                   Enumerable.SequenceEqual(
+                       descriptorX.Attributes.OrderBy(
+                           attribute => TagHelperAttributeDescriptorComparer.Default.GetHashCode(attribute)),
+                       descriptorY.Attributes.OrderBy(
+                           attribute => TagHelperAttributeDescriptorComparer.Default.GetHashCode(attribute)),
+                       TagHelperAttributeDescriptorComparer.Default);
```

Comment:
- `TypeName` takes care of the `Attributes` collection.  do not do this extra comparison, especially because `!Enumerable.SequenceEqual()` here would mean two instances with the same hash code are not equal.
- also remove the `Prefix` from the comparison for similar reasons.